### PR TITLE
ODPM-419 db/sql

### PR DESCRIPTION
**Pull request by jozef-slezak**
Initial commit with sql_broker_api.go & slice_utils.go. This is still work in progress.

### CODE REVIEW

File: db/sql/cassandra/cassa_broker_impl.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+				ptrs = append(ptrs, field.Interface())
+			}
+			//case r.Ptr, r.Interface, r.Array, r.Map, r.SliceIt, r.UnsafePointer, r.Chan /*TODO slice, map...*/ :
+		default:
+			if field.CanAddr() {
+				ptrs = append(ptrs, field.Addr().Interface())
+			} else if field.IsValid() {
+				ptrs = append(ptrs, field.Interface())
+			} else {
+				panic("ivalid field")
```
Comment (lukasmacko) [inline db/sql/cassandra/cassa_broker_impl.go]: "invalid field" typo

File: db/sql/cassandra/cassa_broker_impl.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+			}
+		}
+	}
+
+	return ptrs
+}
+
+// Close the iterator. Not the error is important (may occure during marshalling/un-marshalling)
```
Comment (lukasmacko) [inline db/sql/cassandra/cassa_broker_impl.go]: `NotE` the error - typo

File: db/sql/cassandra/cassa_data_test.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cassandra_test
+
+import (
+	"github.com/willfaught/gockle"
+	"github.com/maraino/go-mock"
+	"errors"
+	reflect2 "github.com/gocassa/gocassa/reflect"
```
Comment (lukasmacko) [inline db/sql/cassandra/cassa_data_test.go]: is this name `reflect2` needed?

File: db/sql/cassandra/examples/simple/Readme.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Simple example
+
+To start examples you have to have Cassandra running on localhost, if you don't have installed it locally
+you can use the following docker image.
+```
+sudo docker run -p 9042:9042 --name cassandra01 -d cassandra:latest
+```
+
+In the example connection to etcd is configured using `--cfg` argument.
```
Comment (lukasmacko) [inline db/sql/cassandra/examples/simple/Readme.md]: cassandra instead of etcd? ...I see there no argument processing

File: Makefile [hunk 5]
```diff
@@ -27,6 +29,7 @@ define test_cover_only
 	@go test -covermode=count -coverprofile=${COVER_DIR}coverage_unit5.out ./utils/addrs
 	@go test -covermode=count -coverprofile=${COVER_DIR}coverage_unit6.out ./core
 	@go test -covermode=count -coverprofile=${COVER_DIR}coverage_unit7.out ./db/keyval/redis
+	@go test -covermode=count -coverprofile=${COVER_DIR}coverage_unit8.out ./db/sql/cassandra
     @echo "# merging coverage results"
     @cd vendor/github.com/wadey/gocovmerge && go install -v
     @gocovmerge ${COVER_DIR}coverage_unit1.out ${COVER_DIR}coverage_unit2.out ${COVER_DIR}coverage_unit3.out ${COVER_DIR}coverage_unit4.out ${COVER_DIR}coverage_unit5.out ${COVER_DIR}coverage_unit6.out ${COVER_DIR}coverage_unit7.out > ${COVER_DIR}coverage.out
```
Comment (lukasmacko) [inline Makefile]: add `${COVER_DIR}coverage_unit8.out` also here

File: db/sql/slice_utils.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	sliceType := sl.Type()
+
+	sliceElemType := sliceType.Elem()
+	sliceElemPtr := sliceElemType.Kind() == reflect.Ptr
+	if sliceElemPtr {
+		sliceElemType = sliceElemType.Elem()
+	}
+	for i := 0; i < 5; i++ {
```
Comment (lukasmacko) [inline db/sql/slice_utils.go]: why is this repeated five times?

File: db/sql/sql_broker_api.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	// Example usage 1 (fill slice by values from iterator):
+	//
+	//    query := sql.SelectFrom(UserTable) + sql.Where(sql.Field(&UserTable.LastName, UserTable, "Bond"))
+	//    iterator := db.ListValues(query)
+	//    users := &[]User{}
+	//    err := sql.SliceIt(users, iterator)
+	//
+	// Example usage 2:
+	//
+	//    query := sql.SelectFrom(UserTable) + "where last_name='Bond'")
```
Comment (lukasmacko) [inline db/sql/sql_broker_api.go]: there is redundant parenthesis at the end of line

File: db/sql/sql_broker_api.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	//
+	// Example usage 3:
+	//
+	//    iterator := db.ListValues("select ID, first_name, last_name from User where last_name='Bond'")
+	//    user := map[string]interface{}
+	//    stop := iterator.GetNext(user)
+	//
+	ListValues(query Expression) ValIterator
+
+	// Delete removes data that from the data store
```
Comment (lukasmacko) [inline db/sql/sql_broker_api.go]: without `that`

File: db/sql/sql_broker_api.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	//
+	//  	 err := db.Exec("CREATE INDEX IF NOT EXISTS...")
+	Exec(statement string) error
+}
+
+// ValIterator is an iterator returned by ListValues call.
+type ValIterator interface {
+	// GetNext retrieves the current "row" from query result. GetValue is un-marshaled into the provided argument.
+	// The stop=true will be returned if there is no more record or if error occurred (to get the error call Close())
+	// Whe the stop=true is returned the outBinding was not updated.
```
Comment (lukasmacko) [inline db/sql/sql_broker_api.go]: When

File: db/keyval/bytes_broker_api.go [hunk 10]
```diff
@@ -27,7 +27,7 @@ type BytesBroker interface {
 	// ListKeys is similar to the ListValues the difference is that values are not fetched
 	ListKeys(prefix string) (BytesKeyIterator, error)
```
Comment (junmipan) [inline db/keyval/bytes_broker_api.go]: Also need option to specify glob matching in applicable methods, such as ListKeys(), ListValues() and Watch().

**Pull request merged**